### DISCUSSION TITLE
[dynamicIO] No longer possible to set dynamic <html> attributes. I18n is broken. Framework solution needed.

### DISCUSSION THREAD

**Discussion by apostolos** [vercel/next.js#71927] [App Router] [answered]
## Next.js v15+ with dynamicIO enabled

After the change to pass all dynamic data (`params`, `searchParams`, `headers`, `cookies`, etc) as `Promise`s, one important use case has not been addressed: the ability to set an `<html>` attribute based on some dynamic data.

Since accessing dynamic data requires a `Suspense` boundary and you cannot add a Suspense boundary on a root layout, there is no way to `await` the dynamic params.

`"use cache"` also doesn't work, since you cannot read dynamic params inside a Page, Layout, or function that is using this directive. Build errors after a timeout with the following message:
> [!CAUTION]
> Filling a cache during prerender timed out like because request specific arguments such as params, searchParams, cookies() or dynamic data was used inside the "use cache".

These are all expected behavior, however one very important use case feels impossible to implement; Internationalization. Basically this entire [documentation page](https://nextjs.org/docs/canary/app/building-your-application/routing/internationalization) is obsolete. The following code is no longer valid:

```js
export async function generateStaticParams() {
  return [{ lang: 'en-US' }, { lang: 'de' }]
}
 
export default function Root({ children, params }) {
  return (
    <html lang={params.lang}>
      <body>{children}</body>
    </html>
  )
}
```

This is very unfortunate. And similarly you cannot set the `dir` attribute dynamically which is a critical feature for RTL content!

## Failed attempts to fix the problem

| Solution | Dev | Build |
| - | - | - |
| Without `Suspense` boundary or `"use cache"` | works | build error |
| Suspense boundary above the `<html>` tag | hydration error | build fails |
| `"use cache"` directive in `layout.tsx` | error | filling cache error |
| `"use cache"` directive inside `<Layout>` | error | filling cache error |
| Reading params synchronously | error | build error |
| Using `use` | works | build error |
| Sync layout, async `<Html params={params}>` | works | build error |
| `<Html>` with `use client` that reads `useParam` | works | build error |
| `AsyncLocalStorage` or accessing Next.js internals | error | build error |
| `useEffect`, `useLayoutEffect`, `useInsertionEffect` | flash of LTR | flash of LTR |
| Injecting lang="" dir="" attributes upstream (e.g. reverse proxy, edge) | I hate my life | yaiks |
 
## Hacky workaround

To fix the `dir="rtl"` situation I ended up using a similar technique that others have for setting classes or attributes for dark mode.

I first set a client cookie in the middleware. Then I read the cookie value and modify the <html> element synchronously, suppressing hydration warnings on the `<html>` tag. All these before the browser has the chance to render anything on the page:

```jsx
export default function Layout({ children }) {
  return (
    <html suppressHydrationWarning={true}>
      <head>
        <script>{langCode}</script>
      </head>
      <body>{children}</body>
    </html>
  );
}
```

Where `langCode`:

```js
const code = `(${() => {
  const cookies = String(document.cookie);
  const pos = cookies.indexOf('__Host-Lang=');
  if (pos > -1) {
    const endPos = cookies.indexOf(';', pos);
    const cookieValue = cookies.slice(pos + 12, endPos > -1 ? endPos : undefined);
    document.documentElement.lang = cookieValue;
    document.documentElement.dir = ['ar', 'fa', 'he', 'ps', 'ur'].includes(cookieValue.slice(0, 2)) ? 'rtl' : 'ltr';
  }
}})();`;

// The code should be injected in the page before any other script so it runs before React hydration
export default code;
```

Unfortunately, this doesn't solve situations when the language depends on outside data (e.g. an article from a CMS). Or if you want to change the language without resorting to full-page navigation.

## Solution needed

I don't know what a React-native or Next.js-native solution for this problem could look like.

To me it feels more of a framework oversight than something that React should address. So, a solution from Next.js seems more appropriate.

Some have suggested adding attributes support in `generateMetadata`, for example: [https://github.com/vercel/next.js/discussions/49415](https://github.com/vercel/next.js/discussions/49415). I guess this could work. Although I think metadata are asynchronously steamed in, so that means that the `<html>` tag is already injected at that point.

Maybe some kind of middleware api could handle all this. I don't know...

The `lang`/`dir` problem might be important enough to deserve a specific solution for it.

**Comment by sebmarkbage** [comment, marked as answer]
The main thing that's currently missing is the flag `dynamicParams = false`. You'll notice that this is disabled in `dynamicIO` because we have plans for a different alternative. Once that's in then you'll be able to await the params without a Suspense boundary since they will all be known statically (from `generateStaticParams`).

(It will also be possible to use a Suspense boundary above html but that's not the best solution since it makes the whole page dynamic.)

**Reply by darthmaim** [reply]
Maybe it would be useful to have a way to do `await params.get('static')` instead of or in addition to `(await params).static`, so static params could just be resolved when using dynamicIO, while dynamic params trigger dynamic rendering.

This would nicely enable things like this:
```tsx
export function Page({ params }) {
  const static = await params.get('static')
  
  return (
    <>
      <div>static: {static}</div>
      <Suspense>
        <DynamicComponent promise={params.get('dynamic')}/>
      <Suspense>
    </>
  )
}

export function generateStaticParams() {
  return [{ static: 'foo' }]
}
```

**Reply by stefanprobst** [reply]
@sebmarkbage can you share more about what the plans for an alternative to `dynamicParams = false` are? i just tried with the new consolidated `cacheComponents` flag and this is still a missing piece in the story.

**Reply by darthmaim** [reply]
@stefanprobst root params (either the old `unstable_rootParams` or the new `next/root-params`) are probably what you are looking for.

**Reply by stefanprobst** [reply]
`dynamicParams=false` redirects to 404 when receiving a value for a path segment which is not returned from `generateStaticParams`. this currently does not work with `dynamicIO`/`cacheComponents`. afaiu `rootParams` solve a different problem.

**Reply by amannn** [reply]
I've just opened https://github.com/vercel/next.js/discussions/84991 to continue the conversation on `dynamicParams = false` for `cacheComponents`.

If you're interested in this, please leave an upvote on the discussion and feel free to join the conversation!

**Comment by daveycodez** [comment]
I had to comment out a bunch of stuff to test this, but I'm correctly able to set dynamic parameters on the html tag. Just put "use cache" at the top of the [locale]/layout.tsx

```tsx
"use cache"

import { hasLocale, Locale } from "next-intl"

import { routing } from "@/i18n/routing"

import { Inter } from "next/font/google"
import { cn } from "@/lib/utils"
import { Providers } from "../providers"
import { notFound } from "next/navigation"
import { getMessages, setRequestLocale } from "next-intl/server"
import NextIntlProvider from "./next-intl-provider"

const inter = Inter({ subsets: ["latin"] })

export default async function LocaleLayout({
    children,
    params
}: {
    children: React.ReactNode
    params: Promise<{ locale: Locale }>
}) {
    const { locale } = await params

    // Ensure that the incoming `locale` is valid
    if (!hasLocale(routing.locales, locale)) {
        notFound()
    }

    // Enable static rendering
    setRequestLocale(locale)

    // Providing all messages to the client
    // side is the easiest way to get started
    const messages = await getMessages({ locale })

    return (
        <html key="html" lang={locale} suppressHydrationWarning>
            <body>
                <NextIntlProvider locale={locale} messages={messages}>
                    <Providers>
                        <div
                            className={cn(inter.className,
                                "flex flex-col w-svw min-h-svh bg-gradient-to-r from-blue-100 to-purple-100 dark:from-blue-950/50 dark:to-purple-950/50"
                            )}
                        >
                            <div className="px-safe pb-safe flex flex-col grow">
                                {children}
                            </div>
                        </div>
                    </Providers>
                </NextIntlProvider>
            </body>
        </html>
    )
}

export async function generateStaticParams() {
    return routing.locales.map((locale) => ({ locale }))
}
```

Only issue here is using any of the next-intl stuff, it complains about using "headers"

I'll keep playing around with dynamicIO

**Reply by daveycodez** [reply]
NextIntlProvider here is a simple "use client" wrapper for NextIntlClientProvider to fix its headers() issue

It totally works but yea still trying some things out across the board.

**Reply by amannn** [reply]
I also did some research with `dynamicIO` and `ppr`, maybe you're interested in this: https://github.com/amannn/next-intl/issues/1493#issuecomment-2476515768

**Comment by apostolos** [comment]
`unstable_rootParams` was added to canary a few weeks ago, so this now works fine. Tested with DIO on or off, works either way. AFAICT, root-level Suspense is not required for this use case. Although that is coming as well, see https://github.com/facebook/react/pull/32069

`./src/[locale]/layout.tsx`:
```tsx
import { unstable_rootParams as rootParams } from 'next/server';

// i18n stuff
import { locales } from '#lang/config.ts';
import { langDir } from '#lang/utils.ts';

export async function generateStaticParams() {
  return locales.map((locale) => ({ locale }));
}

export default async function RootLayout({ children }: React.PropsWithChildren) {
  const { locale } = await rootParams();
  if (typeof locale !== 'string') {
    locale = 'en-GB';
  }
  return (
    <html lang={locale} dir={langDir(locale)}>
      <body>
        {children}
      </body>
    </html>
  );
}
```

**Reply by amannn** [reply]
Does `rootParams` work without any issues for you? I had some findings last time I tried it: https://github.com/amannn/next-intl/discussions/1627

**Reply by apostolos** [reply]
404 pages are broken in non-root layouts in general. It is a known bug and unrelated to rootParams. There are many bugs due to non-root layouts, e.g. intercepted routes with DIO enabled are broken in production.

It is unfortunate that non-root layouts are not getting more love. It's probably due to the fact that Vercel themselves are using a "multi zone" architecture: https://x.com/leeerob/status/1882280042124005615

I haven't run into the other issues mentioned in your post, but that's probably because we manually do hard navigations on language change.

**Reply by daveycodez** [reply]
I turned on dynamicIO for an existing PPR setup and went through a lot of steps to get everything working on dev just to try output: "export" and got the dreaded "server actions are not supported in static export" with no way to trace what caused it. Oof. Gonna try starting fresh from a blank Next install to see if I can locate it

**Reply by daveycodez** [reply]
Turns out "use cache" converts anywhere it is put into a "server action"

IDK how they expect that to be the new default if it can't work with a static export? And I can't "conditionally" render "use cache" for a hybrid app. I'm able to conditionally use PPR by itself without dynamicIO and switch to static export with no issues... For example, can't do this...

```ts
process.env.NEXT_PUBLIC_EXPORT != "true" ? "use cache" : ""
```

The last talk I watched he said PPR will bundle dynamicIO with it :/ I'm not really sure what approach to take here but I do know that PPR by itself in production isn't working properly with client-side navigation to "PPR" pages, it's still firing function calls on link clicks after prefetch even though the dynamic part is in the layout and not the page itself. Lots of fiddling to do...
